python/ray/tune/trial_runner.py

```diff
@@ -84,12 +92,81 @@ def __init__(self,
         self._global_time_limit = float(
             os.environ.get("TRIALRUNNER_WALLTIME_LIMIT", float('inf')))
         self._total_time = 0
+        self._iteration = 0
+        self._verbose = verbose
+        self._queue_trials = queue_trials
+
         self._server = None
         if launch_web_server:
+            self._server_port = server_port
             self._server = TuneServer(self, server_port)
+
+        self._trials = []
         self._stop_queue = []
-        self._verbose = verbose
-        self._queue_trials = queue_trials
+        self._trial_checkpoints = {}
+        self._checkpoint_dir = checkpoint_dir
+        self._checkpoint_mode = checkpoint_mode
+
+    def checkpoint(self):
+        """Saves execution state to `self._checkpoint_dir` if provided."""
+        if not self._checkpoint_mode or not self._checkpoint_dir:
+            return
+        checkpoint_dir = self._checkpoint_dir
+        if not os.path.exists(checkpoint_dir):
+            logger.debug("Checkpoint directory newly created.")
+            logger.warning("Search Algorithm and Scheduler not checkpointed.")
+            os.makedirs(checkpoint_dir)
+        runner_state = {
+            "checkpoints": list(
+                self.trial_executor.get_checkpoints().values()),
+            "runner": self
+        }
+        tmp_file_name = os.path.join(checkpoint_dir, ".tmp_checkpoint")
+        with open(tmp_file_name, "wb") as f:
+            cloudpickle.dump(runner_state, f)
```

Comment:
Consider storing the state in plain JSON, with only the non-jsonable parts in pickle format. This makes it easier to debug when things are going wrong.

Reply:
Done.